sched/group/group_setegid.c

```diff
@@ -0,0 +1,84 @@
+/****************************************************************************
+ * sched/group/group_setegid.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <unistd.h>
+#include <assert.h>
+#include <errno.h>
+
+#include <sched/sched.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: setegid
+ *
+ * Description:
+ *   The setegid() function sets the effective group ID of the calling
+ *   process to gid, given appropriate privileges.
+ *
+ * Input Parameters:
+ *   gid - Identity to set the various process's group ID attributes to.
+ *
+ * Returned Value:
+ *   Zero if successful and -1 in case of failure, in which case errno is set
+ *   to one of he following values:
+ *
+ *   EINVAL - The value of the uid argument is invalid and not supported by
+ *            the implementation.
+ *   EPERM  - The process does not have appropriate privileges and uid does
+ *            not match the effective group ID or the saved set-group-ID.
+ *
+ ****************************************************************************/
+
+int setegid(gid_t gid)
+{
+  FAR struct tcb_s *rtcb;
+  FAR struct task_group_s *rgroup;
+
+  /* Verify that the GID is in the valid range of 0 through INT16_MAX.
+   * OpenGroup.org does not specify a GID_MAX or GID_MIN.  Instead we use a
+   * priori knowledge that gid_t is type int16_t.
+   */
+
+  if ((uint16_t)gid > INT16_MAX)
+    {
+      set_errno(EINVAL);
+      return ERROR;
+    }
+
+  /* Get the currently executing thread's task group. */
+
+  rtcb   = this_task();
+  rgroup = rtcb->group;
+
+  /* Set the task group's group identity. */
+
+  DEBUGASSERT(rgroup != NULL);
+  rgroup->tg_egid = gid;
+  return OK;
```

Comment:
Eventually, this needs to be fixed. And unprivileged task should not be permitted to change its effective GID.

Reply:
> > Yes, it's hard to support SUID/SGID without binfs because posix_spawn doesn't have parameters to pass the custom uid/gid.
> 
> They are available in the builtin registry data, that is how they are obtained in the binfs case.

it's simple toget this info in nsh, but how we pass this info to kernel is hard without modifying task_create or task_spawn.

Reply:
It's the independent but related job:

1. Assign the task with the correct uid/gid from user login or SUID/SGID
2. Do the permission check in various place

Item 2 need the work from item1, but item 2 need more work. We need do the huge work step by step, otherwise nothing will happen.